keps/sig-node/3288-separate-stdout-from-stderr/README.md

```diff
@@ -0,0 +1,817 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently, kubelet actually [has the potential](https://github.com/kubernetes/kubernetes/blob/69cabe778bccfc5f4fa9b6788c9def005e7d959d/pkg/kubelet/kuberuntime/logs/logs.go#L281) 
+to return certain log stream of a container, but this ability is not exposed to the user.
+This proposal enables users to view certain log stream of a container,
+and aims to extend kubelet and api-server with the ability to return certain
+container log stream.
+
+## Motivation
+
+Users could fetch logs of the container, but now kubelet always returns
+combined stdout and stderr logs, which is not convenient for users who only want to
+view certain log stream. 
+Meanwhile, [many users](https://github.com/kubernetes/kubernetes/issues/28167) have been interested in the ability to 
+retrieve certain log stream of a container, so it is great to implement this long-wanted feature.
+
+### Goals
+
+- Enable api-server to return certain log stream of a container
+- Enable users to fetch certain log stream of a container
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+Add a new field `Stream` to `PodLogOptions` to allow users to indicate which log stream
+they want to retrieve. To maintain backward compatibility, if this field is not set,
+the combined stdout and stderr from the container will be returned to users.
+
+### User Stories
+
+Bob runs a service "foo" that continuously prints informational messages to stdout,
+and prints warnings/errors to stderr. When the service "foo" is misbehaving, Bob wants
+to know what is going on as soon as possible, so he checks the logs of the service "foo",
+but the error message is not easy to notice because the apiserver always returns the combined
+stdout and stderr from the container.
+
+This problem could be resolved if Bob is able to specify that he only wants stderr.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+
+## Design Details
+
+### Changes of kube-apiserver
+
+Add a new field `Stream` to `k8s.io/kubernetes/pkg/apis/core.PodLogOptions`:
+
+```go
+// PodLogOptions is the query options for a Pod's logs REST call
+type PodLogOptions struct {
+...
+    // If set, return the given log stream of the container.
+	// Otherwise, the combined stdout and stderr from the container is returned.
+	// Available values are: "stdout", "stderr", "" (empty string).
+    Stream string
```

Comment:
is it a string for implementation simplicity or expect any other values to be added later? Perhaps a custom type with three constants instead?

Reply:
> Perhaps a custom type with three constants instead?

Yeah I think this is better, I have updated the proposal.